### PR TITLE
feat: Enhancements in the way how OAuth2 providers can be configured

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
Enhancements in the way how OAuth2 providers can be configured

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/20496

### How to test this PR?

1. Deploy Eclipse Che
2. Create secrets:
```
  kubectl apply -f - <<EOF
kind: Secret
apiVersion: v1
metadata:
  name: che-gitlab
  namespace: eclipse-che
  labels:
    app.kubernetes.io/component: oauth-scm-configuration
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:
    che.eclipse.org/oauth-scm-server: gitlab
    che.eclipse.org/scm-server-endpoint: <endpoint>
data:
  id: aWQ=
  secret: c2VjcmV0
type: Opaque
EOF
```
```
  kubectl apply -f - <<EOF
kind: Secret
apiVersion: v1
metadata:
  name: che-github
  namespace: eclipse-che
  labels:
    app.kubernetes.io/component: oauth-scm-configuration
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:
    che.eclipse.org/oauth-scm-server: github
    che.eclipse.org/scm-server-endpoint: <endpoint>
data:
  id: aWQ=
  secret: c2VjcmV0
type: Opaque
EOF
```

3. Checks that che-server pod contains the following files
```
CHE_POD=$(kubectl get pods -n eclipse-che  | awk '{print $1}' | grep -vE 'che-dashboard|che-operator' | grep che)
kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "printenv CHE_INTEGRATION_GITLAB_SERVER__ENDPOINTS"
kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "printenv CHE_INTEGRATION_GITHUB_SERVER__ENDPOINTS"

kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "cat  /che-conf/oauth/github/id"
kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "cat  /che-conf/oauth/github/secret"
kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "cat  /che-conf/oauth/gitlab/id"
kubectl exec -it $CHE_POD -n eclipse-che  -- bash  -c "cat  /che-conf/oauth/gitlab/secret"
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/server/server_deployment.go [hunk 1]
```diff
@@ -420,3 +429,97 @@ func MountBitBucketOAuthConfig(deployContext *deploy.DeployContext, deployment *
...
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: secrets[0].Name,
+					},
+				},
+			})
+		container.VolumeMounts = append(container.VolumeMounts,
+			corev1.VolumeMount{
+				Name:      secrets[0].Name,
+				MountPath: deploy.GitHubOAuthConfigMountPath,
+			})
```
Comment (mmorhun) [inline pkg/deploy/server/server_deployment.go:461]: Optionally, you may move this to a separate function

File: pkg/deploy/defaults.go [hunk 2]
```diff
@@ -156,6 +156,14 @@ const (
...
 	BitBucketOAuthConfigPrivateKey  = "private.key"
 	BitBucketOAuthConfigConsumerKey = "consumer.key"
+
+	GitHubOAuthConfigMountPath    = "/che-conf/oauth/github"
+	GitHubOAuthConfigClientId     = "id"
+	GitHubOAuthConfigClientSecret = "secret"
+
+	GitLabOAuthConfigMountPath    = "/che-conf/oauth/gitlab"
+	GitLabOAuthConfigClientId     = "id"
+	GitLabOAuthConfigClientSecret = "secret"
```
Comment (mmorhun) [inline pkg/deploy/defaults.go:166]: These are not ClientId and ClientSecret values, but sort of keys. Please name accordingly.
The same for GitHub.

**Review by mmorhun — APPROVED** [review]

File: pkg/deploy/defaults.go (hunk 2 above)
Comment (tolusha) [reply pkg/deploy/defaults.go:166]: Unfortunately it is impossible. When secret is mounted into a container then files names = keys names

File: pkg/deploy/server/server_configmap.go [hunk 3]
```diff
@@ -308,15 +308,24 @@ func (s *Server) getCheConfigMapData() (cheEnv map[string]string, err error) {
 
 	addMap(cheEnv, s.deployContext.CheCluster.Spec.Server.CustomCheProperties)
 
-	// Update BitBucket endpoints
-	secrets, err := deploy.GetSecrets(s.deployContext, map[string]string{
+	err = setBitBucketEndpoints(s.deployContext, cheEnv)
```
Comment (sleshchenko) [inline pkg/deploy/server/server_configmap.go:311]:
```suggestion
	err = setBitbucketEndpoints(s.deployContext, cheEnv)
```
let's be consistent with the way bitbucket names themselves.

**Pull request merged**